documentation/general/dotnet-run-file-multi-file.md

```diff
@@ -0,0 +1,136 @@
+- Performance from CLI and IDE point of view
+- Workspace context in IDE for a given file (does each entry-point file map to a project? or do they all map to the same project? or do they all map to different projects?)
+
+## Approaches
+
+### Adopt default items according to the SDK being used
+
+- Defaults all files to work like a regular project
+- Can be disabled for a given entry-point file by adding a directive like `#:property EnableDefaultItems=false`
+- Can be disbled for a whole directory by adding a file `Directory.Build.props` with the following content:
+
+    ```xml
+    <Project>
+      <PropertyGroup>
+        <EnableDefaultItems>false</EnableDefaultItems>
+      </PropertyGroup>
+    </Project>
+    ```
+
+- Works for all SDKs including those that add other default items beyond **.cs* files
+- Maps naturally to converted project
+- How to know when to stop including files in the directory tree? Is this already implicitly handled given existing project-based default items behavior?
+- How to deal with multiple entry-point files?
+  - Include all entry-point files, no special treatment at the file level
+  - Update compiler to handle multiple top-level statements entry-point files by new option that changes generated type name to one derived from the file name, e.g. `app` or `Program_app.cs` for `app.cs` and `app2` or `Program_app2` for `app2.cs`, so that type name conflicts are avoided
+  - `dotnet run app.cs` will set the [build/compiler options](https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-options/advanced#mainentrypoint-or-startupobject) to use the entry-point type that corresponds to the entry-point file that was passed
+- With compiler changes, enables all scenarios mentioned above
+
+### Require explicit opt-in to enable default items
+
+- Defaults all entry-point files to be single-file apps
+- Require default items to be explicitly enabled via the entry-point file
+- Can be enabled for a given entry-point file by adding a directive like `#:property EnableDefaultItems=true`
+- Works for all SDKs including those that add other default items beyond **.cs* files once opted-in
+- Doesn't map to the natural default for coverted projects, but maybe that's OK? When converted a new directory is created and only the relevant files are brought in based on whether default items were enabled.
+- Likely hard to discover for users. Mitigation could be to have a special directive, e.g. `#:multifile`
+
+### Support directives to specify included files
+
+- Defaults all entry-point files to be single-file apps
+- Support directives to specify included files patterns, e.g. `#:include *.cs` or `#:include **/*.cs`
+  - This seems pretty good at first but is limited to `Compile` items only. Maybe has utility as a limited option to go along with `#:property EnableDefaultItems=false` for folks who want to control included **.cs* files only.
```

Comment:
The other issue is that it means _more work_ to have a single file app look like the `dotnet new console` experience.